x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/CreateEnrollmentToken.java

```diff
@@ -0,0 +1,231 @@
+    }
+
+    protected URL createAPIKeyUrl() throws MalformedURLException, URISyntaxException {
+        return new URL(defaultUrl, (defaultUrl.toURI().getPath() + "/_security/api_key").replaceAll("/+", "/"));
+    }
+
+    protected URL getHttpInfoUrl() throws MalformedURLException, URISyntaxException {
+        return new URL(defaultUrl, (defaultUrl.toURI().getPath() + "/_nodes/_local/http").replaceAll("/+", "/"));
+    }
+
+    protected static List<String> getBoundAddresses(Map<?, ?> nodesInfo) {
+        nodesInfo = (Map<?, ?>) nodesInfo.get("nodes");
+        Map<?, ?> nodeInfo = (Map<?, ?>) nodesInfo.values().iterator().next();
+        Map<?, ?> http = (Map<?, ?>) nodeInfo.get("http");
+        return (ArrayList<String>) http.get("bound_address");
+    }
+
+    static String getVersion(Map<?, ?> nodesInfo) {
+        nodesInfo = (Map<?, ?>) nodesInfo.get("nodes");
+        Map<?, ?> nodeInfo = (Map<?, ?>) nodesInfo.values().iterator().next();
+        return nodeInfo.get("version").toString();
+    }
+
+    protected String getApiKeyCredentials(String user, SecureString password, String action) throws Exception {
+        final CheckedSupplier<String, Exception> createApiKeyRequestBodySupplier = () -> {
+            XContentBuilder xContentBuilder = JsonXContent.contentBuilder();
+            xContentBuilder.startObject()
+                .field("name", "enrollment_token_API_key_" + UUIDs.base64UUID())
+                .field("expiration", ENROLL_API_KEY_EXPIRATION)
+                .startObject("role_descriptors")
+                .startObject("create_enrollment_token")
+                .array("cluster", action)
+                .endObject()
+                .endObject()
+                .endObject();
+            return Strings.toString(xContentBuilder);
+        };
+
+        final URL createApiKeyUrl = createAPIKeyUrl();
+        final HttpResponse httpResponseApiKey = client.execute("POST", createApiKeyUrl, user, password,
+            createApiKeyRequestBodySupplier, is -> responseBuilder(is));
+        final int httpCode = httpResponseApiKey.getHttpStatus();
+
+        if (httpCode != HttpURLConnection.HTTP_OK) {
+            logger.error("Error " + httpCode + "when calling GET " + createApiKeyUrl + ". ResponseBody: " +
+                httpResponseApiKey.getResponseBody());
+            throw new IllegalStateException("Unexpected response code [" + httpCode + "] from calling POST "
+                + createApiKeyUrl);
+        }
+
+        final String apiKey = Objects.toString(httpResponseApiKey.getResponseBody().get("api_key"), "");
+        final String apiId = Objects.toString(httpResponseApiKey.getResponseBody().get("id"), "");
+        if (Strings.isNullOrEmpty(apiKey) || Strings.isNullOrEmpty(apiId)) {
+            throw new IllegalStateException("Could not create an api key.");
+        }
+        return Base64.getEncoder().encodeToString((apiId + ":" + apiKey).getBytes(StandardCharsets.UTF_8));
```

Comment:
Missed that in the review @BigPandaToo . I don;t think there is need to base64 encode the API key before we put it as a value in the token, as the token itself will be Base64 encoded. 

I think the length gains are more significant than the additiional effort the consumers of the token need to make to base64 the string before using it in the Authorization header and this behavior is also consistent with the create api key API . 

Can you please tackle this change in a short follow up PR ?

Reply:
Right, forgot that it won't be used directly